pallets/capacity/src/lib.rs

```diff
@@ -513,24 +516,86 @@ impl<T: Config> Pallet<T> {
 		unstaker: &T::AccountId,
 		amount: BalanceOf<T>,
 	) -> Result<BalanceOf<T>, DispatchError> {
-		let mut staking_account =
-			Self::get_staking_account_for(unstaker).ok_or(Error::<T>::NotAStakingAccount)?;
-		ensure!(amount <= staking_account.active, Error::<T>::AmountToUnstakeExceedsAmountStaked);
+		StakingAccountLedger::<T>::try_mutate_exists(
+			unstaker,
+			|maybe_staking_account| -> Result<BalanceOf<T>, DispatchError> {
+				let mut staking_account =
+					maybe_staking_account.take().ok_or(Error::<T>::NotAStakingAccount)?;
+				ensure!(
+					amount <= staking_account.active,
+					Error::<T>::AmountToUnstakeExceedsAmountStaked
+				);
+
+				let actual_unstaked_amount = staking_account.withdraw(amount)?;
+				if staking_account.active.is_zero() {
+					*maybe_staking_account = None;
+				} else {
+					*maybe_staking_account = Some(staking_account);
+				}
+				Ok(actual_unstaked_amount)
+			},
+		)
+	}
 
+	fn add_unlock_chunk(
+		unstaker: &T::AccountId,
+		actual_unstaked_amount: BalanceOf<T>,
+	) -> Result<(), DispatchError> {
 		let current_epoch: T::EpochNumber = Self::get_current_epoch();
 		let thaw_at =
 			current_epoch.saturating_add(T::EpochNumber::from(T::UnstakingThawPeriod::get()));
+		let mut unlocks = Self::get_unstake_unlocking_for(unstaker).unwrap_or_default();
 
-		let unstake_result = staking_account.withdraw(amount, thaw_at)?;
+		let unlock_chunk: UnlockChunk<BalanceOf<T>, T::EpochNumber> =
+			UnlockChunk { value: actual_unstaked_amount, thaw_at };
+		unlocks
+			.try_push(unlock_chunk)
+			.map_err(|_| Error::<T>::MaxUnlockingChunksExceeded)?;
 
-		Self::set_staking_account(&unstaker, &staking_account);
+		UnstakeUnlocks::<T>::set(unstaker, Some(unlocks));
+		Ok(())
+	}
 
-		Ok(unstake_result)
+	// Calculates a stakable amount from a proposed amount.
+	pub(crate) fn get_stakable_amount_for(
+		staker: &T::AccountId,
+		proposed_amount: BalanceOf<T>,
+	) -> BalanceOf<T> {
+		let account_balance = T::Currency::free_balance(&staker);
+		account_balance
+			.saturating_sub(T::MinimumTokenBalance::get())
+			.min(proposed_amount)
+	}
+
+	pub(crate) fn do_withdraw_unstaked(
+		staker: &T::AccountId,
+	) -> Result<BalanceOf<T>, DispatchError> {
+		let current_epoch = Self::get_current_epoch();
+		let mut total_unlocking: BalanceOf<T> = Zero::zero();
+
+		let mut unlocks =
+			Self::get_unstake_unlocking_for(staker).ok_or(Error::<T>::NoUnstakedTokensAvailable)?;
+		let amount_withdrawn = unlock_chunks_reap_thawed::<T>(&mut unlocks, current_epoch);
+		ensure!(!amount_withdrawn.is_zero(), Error::<T>::NoThawedTokenAvailable);
+		if unlocks.is_empty() {
+			UnstakeUnlocks::<T>::set(staker, None);
+		} else {
+			total_unlocking = unlock_chunks_total::<T>(&unlocks);
+			UnstakeUnlocks::<T>::set(staker, Some(unlocks));
+		}
+		let staking_account = Self::get_staking_account_for(staker).unwrap_or_default();
+		let total_locked = staking_account.active.saturating_add(total_unlocking);
+		if total_locked.is_zero() {
```

Comment:
I think we should keep the set delete method interface

```
delete_staking_account
set_staking_account
```

I am concern that someone is going to use these directly at some point and mess up the accounting.

Reply:
I think encapsulating one-liner functions that are called only from one place is not a good way to prevent others from making calls they should not make.  If anything it communicates that the function is there to be called from other places.

Reply:
Ok can we please wrap these in their own functions. It will make it easier for updating for the new fungible trait.

Reply:
ah ok, I will do a follow-up then.